src/libos/src/vm/vm_clean.rs

```diff
@@ -0,0 +1,179 @@
+            return vec![Self {
+                range,
+                inner: Inner::Multi(chunk_manager_ref.unwrap()),
+            }];
+        }
+
+        Self::split_req(range, chunk_manager_ref)
+    }
+
+    fn inner(&self) -> &Inner {
+        &self.inner
+    }
+
+    fn split_req(range: VMRange, chunk_manager: Option<Arc<SgxMutex<ChunkManager>>>) -> Vec<Self> {
+        let inner = if chunk_manager.is_some() {
+            Inner::Multi(chunk_manager.unwrap())
+        } else {
+            Inner::Single
+        };
+
+        let num = range.size() / REQ_MAX_SIZE;
+        let reqs = (0..num)
+            .into_iter()
+            .map(|i| {
+                let range_start = range.start() + i * REQ_MAX_SIZE;
+                let range_end = if i < num - 1 {
+                    range_start + REQ_MAX_SIZE
+                } else {
+                    range.end()
+                };
+                Self {
+                    // Just split to small ranges, safe
+                    range: unsafe { VMRange::from_unchecked(range_start, range_end) },
+                    inner: inner.clone(),
+                }
+            })
+            .collect::<Vec<CleanReq>>();
+
+        reqs
+    }
+}
+
+// These requests are either send to clean queue or clean by current thread
+pub fn send_reqs(mut reqs: Vec<CleanReq>) -> Result<()> {
+    if CLEAN_REQ_QUEUE.len() > *CLEAN_QUE_HIGH_WATERMARK && reqs.len() <= REQ_NUM_THREASHOLD {
+        // There are too many cleaning requests in the queue, and since the new requests are not too big,
+        // return and let the current thread do the cleaning.
+        return_errno!(ENOMEM, "there are too many cleaning requests");
+    }
+
+    while !reqs.is_empty() {
+        let req = reqs.pop().unwrap();
+        CLEAN_REQ_QUEUE.send(req);
+        // Wake one cleaning thread
+        CLEAN_THREAD_WAITER_QUEUE.wake_one();
+    }
+    Ok(())
+}
+
+pub fn init_vm_clean_thread() -> Result<()> {
+    (0..*CLEAN_THREAD_WORKER_NUM).into_iter().for_each(|_| {
+        async_rt::task::spawn(async {
+            mem_worker_thread_func().await;
+        });
+    });
+    Ok(())
+}
+
+async fn mem_worker_thread_func() {
+    let mut waiter = Waiter::new();
+    CLEAN_THREAD_WAITER_QUEUE.enqueue(&mut waiter);
+    loop {
+        if let Ok(req) = CLEAN_RUNNER.try_recv() {
+            clean_and_return_dirty_range(req).expect("clean and return dirty range error");
+        } else {
+            waiter.reset();
+            waiter.wait().await;
+        }
+    }
+
+    // This never reaches
+    unreachable!();
+}
+
+// This function is called by non-cleaning thread to help clean the dirty range
+// when the libos exits or the user thread fails to allocate memory
+pub fn help_dirty_range_clean() {
+    while let Ok(req) = CLEAN_RUNNER.try_recv() {
+        clean_and_return_dirty_range(req).expect("clean and return dirty range error");
+    }
+}
+
+fn clean_and_return_dirty_range(req: CleanReq) -> Result<()> {
```

Comment:
The name confused me at first. The name of this function seems that it would return a dirty range. Actually it means "clean the dirty range then return it back". This function needs a better name.

Reply:
Rename this function. Thank you for your review.